usb_cam silently accepts `pixel_format:=uyvy` / yuv422, even when not supported by the device, and sends images with the wrong `encoding` value

(tested on Iron, built from source on latest usb_cam ros2 branch, Ubuntu 22.04 in a VM with macOS arm64 host)

With default settings or `pixel_format:=yuyv`, the resulting Image messages have `encoding: yuv422_yuy2`. RViz and Foxglove Studio can display these correctly:

<img width="361" alt="image" src="https://github.com/ros-drivers/usb_cam/assets/14237/565b2d44-962a-4a18-8616-61c5b5156ce7">

<img width="894" alt="image" src="https://github.com/ros-drivers/usb_cam/assets/14237/917dcc2e-9c3a-4557-a875-6dc7d04555be">


But when I set `pixel_format:=uyvy`, the Image has `encoding: yuv422`, and it looks like this:

<img width="358" alt="image" src="https://github.com/ros-drivers/usb_cam/assets/14237/91fbe876-ed4c-4c51-b166-37fbfff957d2">

<img width="803" alt="image" src="https://github.com/ros-drivers/usb_cam/assets/14237/231e91d7-d102-4638-a3a9-f56c6632e8ed">


It seems like maybe what is happening is that the pixel data does not actually change between `pixel_format:=yuyv` and `pixel_format:=uyvy`, but only the image message's `encoding` changes.

Sorry for the delay, I was on vacation.

> does your device support capturing in both formats? This driver should print to the console what formats your device supports. Can you share it here?

No, I guess it doesn't. This is the full output:

```
jacob@parallels-jammy:~/Desktop/ros2_ws$ ros2 run usb_cam usb_cam_node_exe --ros-args -p pixel_format:=uyvy -p video_device:=/dev/video2
[INFO] [1692056422.608974546] [usb_cam]: camera_name value: default_cam
[WARN] [1692056422.609487650] [usb_cam]: framerate: 30.000000
[INFO] [1692056422.611439995] [usb_cam]: using default calibration URL
[INFO] [1692056422.612138243] [usb_cam]: camera calibration URL: file:///home/jacob/.ros/camera_info/default_cam.yaml
[ERROR] [1692056422.612228774] [camera_calibration_parsers]: Unable to open camera calibration file [/home/jacob/.ros/camera_info/default_cam.yaml]
[WARN] [1692056422.612362466] [usb_cam]: Camera calibration file /home/jacob/.ros/camera_info/default_cam.yaml not found
[INFO] [1692056422.612402961] [usb_cam]: Starting 'default_cam' (/dev/video2) at 640x480 via mmap (uyvy) at 30 FPS
[INFO] [1692056422.680840598] [usb_cam]: This devices supproted formats:
[INFO] [1692056422.680983414] [usb_cam]: 	YUYV 4:2:2: 1024 x 576 (15 Hz)
[INFO] [1692056422.680990079] [usb_cam]: 	YUYV 4:2:2: 1280 x 720 (10 Hz)
[INFO] [1692056422.680992996] [usb_cam]: 	YUYV 4:2:2: 1600 x 896 (7 Hz)
[INFO] [1692056422.680995829] [usb_cam]: 	YUYV 4:2:2: 160 x 120 (30 Hz)
[INFO] [1692056422.680998412] [usb_cam]: 	YUYV 4:2:2: 160 x 90 (30 Hz)
[INFO] [1692056422.681001078] [usb_cam]: 	YUYV 4:2:2: 176 x 144 (30 Hz)
[INFO] [1692056422.681003494] [usb_cam]: 	YUYV 4:2:2: 320 x 180 (30 Hz)
[INFO] [1692056422.681006119] [usb_cam]: 	YUYV 4:2:2: 320 x 240 (30 Hz)
[INFO] [1692056422.681008786] [usb_cam]: 	YUYV 4:2:2: 352 x 288 (30 Hz)
[INFO] [1692056422.681011244] [usb_cam]: 	YUYV 4:2:2: 432 x 240 (30 Hz)
[INFO] [1692056422.681014368] [usb_cam]: 	YUYV 4:2:2: 640 x 480 (30 Hz)
[INFO] [1692056422.681016868] [usb_cam]: 	YUYV 4:2:2: 800 x 448 (30 Hz)
[INFO] [1692056422.681019409] [usb_cam]: 	YUYV 4:2:2: 800 x 600 (24 Hz)
[INFO] [1692056422.681021909] [usb_cam]: 	YUYV 4:2:2: 864 x 480 (24 Hz)
[INFO] [1692056422.681024409] [usb_cam]: 	YUYV 4:2:2: 960 x 720 (15 Hz)
[INFO] [1692056422.681026992] [usb_cam]: 	Y/CbCr 4:2:0: 1024 x 576 (30 Hz)
[INFO] [1692056422.681030491] [usb_cam]: 	Y/CbCr 4:2:0: 1280 x 720 (30 Hz)
[INFO] [1692056422.681032908] [usb_cam]: 	Y/CbCr 4:2:0: 1600 x 896 (30 Hz)
[INFO] [1692056422.681035324] [usb_cam]: 	Y/CbCr 4:2:0: 160 x 120 (30 Hz)
[INFO] [1692056422.681037949] [usb_cam]: 	Y/CbCr 4:2:0: 160 x 90 (30 Hz)
[INFO] [1692056422.681040365] [usb_cam]: 	Y/CbCr 4:2:0: 176 x 144 (30 Hz)
[INFO] [1692056422.681042823] [usb_cam]: 	Y/CbCr 4:2:0: 1920 x 1080 (30 Hz)
[INFO] [1692056422.681045573] [usb_cam]: 	Y/CbCr 4:2:0: 320 x 180 (30 Hz)
[INFO] [1692056422.681047947] [usb_cam]: 	Y/CbCr 4:2:0: 320 x 240 (30 Hz)
[INFO] [1692056422.681050489] [usb_cam]: 	Y/CbCr 4:2:0: 352 x 288 (30 Hz)
[INFO] [1692056422.681052863] [usb_cam]: 	Y/CbCr 4:2:0: 432 x 240 (30 Hz)
[INFO] [1692056422.681055238] [usb_cam]: 	Y/CbCr 4:2:0: 640 x 360 (30 Hz)
[INFO] [1692056422.681057571] [usb_cam]: 	Y/CbCr 4:2:0: 640 x 480 (30 Hz)
[INFO] [1692056422.681060446] [usb_cam]: 	Y/CbCr 4:2:0: 800 x 448 (30 Hz)
[INFO] [1692056422.681062946] [usb_cam]: 	Y/CbCr 4:2:0: 800 x 600 (30 Hz)
[INFO] [1692056422.681065320] [usb_cam]: 	Y/CbCr 4:2:0: 864 x 480 (30 Hz)
[INFO] [1692056422.681067945] [usb_cam]: 	Y/CbCr 4:2:0: 960 x 720 (30 Hz)
[INFO] [1692056422.681071361] [usb_cam]: Setting 'brightness' to 50
unknown control 'brightness'

unknown control 'white_balance_temperature_auto'

[INFO] [1692056422.690110048] [usb_cam]: Setting 'white_balance_temperature_auto' to 1
[INFO] [1692056422.690173541] [usb_cam]: Setting 'exposure_auto' to 3
unknown control 'exposure_auto'

[INFO] [1692056422.696893678] [usb_cam]: Setting 'focus_auto' to 0
unknown control 'focus_auto'

[INFO] [1692056422.704036013] [usb_cam]: Timer triggering every 33 ms
^C[INFO] [1692056439.905122082] [rclcpp]: signal_handler(signum=2)
[WARN] [1692056439.905484797] [usb_cam]: Shutting down
[ros2run]: Segmentation fault
```

But if a format is not supported by the device, then the usb_cam node probably shouldn't silently pretend like it was (sending the yuyv image data in a message that claims to have `encoding: uyvy`). It should instead either throw an error and refuse to launch with the invalid configuration, or it should fall back to a supported encoding (but report the correct encoding in the image message's `encoding` field).

--- a/config/params_2.yaml
+++ b/config/params_2.yaml
@@ -4,8 +4,8 @@
       framerate: 15.0
       io_method: "mmap"
       frame_id: "camera2"
-      pixel_format: "mjpeg2rgb"  # see usb_cam/supported_formats for list of supported formats
-            av_device_format: "YUV422P"
+      pixel_format: "mjpeg2rgb"
+      av_device_format: "YUV422P"
       image_width: 640
       image_height: 480
       camera_name: "test_camera2"

--- a/include/usb_cam/formats/av_pixel_format_helper.hpp
+++ b/include/usb_cam/formats/av_pixel_format_helper.hpp
@@ -960,6 +960,7 @@ inline AVPixelFormat get_av_pixel_format_from_string(const std::string & str)
   return STR_2_AVPIXFMT.find(fullFmtStr)->second;
 }
 
+
 /// @brief Get ROS PixelFormat from AVPixelFormat.
 /// @param avPixelFormat AVPixelFormat
 /// @return String specifying the ROS pixel format.
@@ -1032,6 +1033,17 @@ inline std::string get_ros_pixel_format_from_av_format(const AVPixelFormat & avP
   return ros_format;
 }
 
+
+/// @brief Overload to support av pixel formats being passed as strings
+/// @param avPixelFormat AvPixelFormat as string
+/// @return String specifying the ROS pixel format.
+inline std::string get_ros_pixel_format_from_av_format(const std::string avPixelFormatStr)
+{
+  auto fmt = get_av_pixel_format_from_string(avPixelFormatStr);
+  return get_ros_pixel_format_from_av_format(fmt);
+}
+
+
 /// @brief Get the number of channels from AVPixelFormat.
 /// @param avPixelFormat AVPixelFormat
 /// @return Number of channels as uint8
@@ -1074,6 +1086,17 @@ inline uint8_t get_channels_from_av_format(const AVPixelFormat & avPixelFormat)
   return channels;
 }
 
+
+/// @brief Overload of function below to support av pixel formats as strings
+/// @param avPixelFormatStr AvPixelFormat as string
+/// @return Number of channels as uint8_t
+inline uint8_t get_channels_from_av_format(const std::string & avPixelFormatStr)
+{
+  auto fmt = get_av_pixel_format_from_string(avPixelFormatStr);
+  return get_channels_from_av_format(fmt);
+}
+
+
 /// @brief Get the pixel bit depth from AVPixelFormat.
 /// @param avPixelFormat AVPixelFormat
 /// @return Bit depth as uint8
@@ -1104,6 +1127,16 @@ inline uint8_t get_bit_depth_from_av_format(const AVPixelFormat & avPixelFormat)
   }
 
   return bit_depth;
+}
+
+
+/// @brief Overload of function below to support passing av pixel formats as strings
+/// @param avPixelFormatStr AVPixelFormat as string
+/// @return Bit depth as uint8
+inline uint8_t get_bit_depth_from_av_format(const std::string & avPixelFormatStr)
+{
+  auto fmt = get_av_pixel_format_from_string(avPixelFormatStr);
+  return get_bit_depth_from_av_format(fmt);
 }
 
 }  // namespace formats

--- a/include/usb_cam/formats/m420.hpp
+++ b/include/usb_cam/formats/m420.hpp
@@ -46,7 +46,7 @@ namespace formats
 class M4202RGB : public pixel_format_base
 {
 public:
-  M4202RGB(const int & width, const int & height)
+  explicit M4202RGB(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "m4202rgb",
       V4L2_PIX_FMT_M420,
@@ -54,8 +54,8 @@ public:
       3,
       8,
       true),
-    m_width(width),
-    m_height(height)
+    m_width(args.width),
+    m_height(args.height)
   {}
 
   /// @brief Convert a YUV420 (aka M420) image to RGB8

--- a/include/usb_cam/formats/mjpeg.hpp
+++ b/include/usb_cam/formats/mjpeg.hpp
@@ -63,22 +63,21 @@ namespace formats
 class RAW_MJPEG : public pixel_format_base
 {
 public:
-  explicit RAW_MJPEG(const AVPixelFormat & avDeviceFormat)
+  explicit RAW_MJPEG(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "raw_mjpeg",
       V4L2_PIX_FMT_MJPEG,
-      get_ros_pixel_format_from_av_format(avDeviceFormat),
-      get_channels_from_av_format(avDeviceFormat),
-      get_bit_depth_from_av_format(avDeviceFormat),
+      get_ros_pixel_format_from_av_format(args.av_device_format_str),
+      get_channels_from_av_format(args.av_device_format_str),
+      get_bit_depth_from_av_format(args.av_device_format_str),
       false)
-  {
-  }
+  {}
 };
 
 class MJPEG2RGB : public pixel_format_base
 {
 public:
-  MJPEG2RGB(const int & width, const int & height, const AVPixelFormat & avDeviceFormat)
+  explicit MJPEG2RGB(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "mjpeg2rgb",
       V4L2_PIX_FMT_MJPEG,
@@ -103,18 +102,18 @@ public:
 
     m_avcodec_context = avcodec_alloc_context3(m_avcodec);
 
-    m_avframe_device->width = width;
-    m_avframe_device->height = height;
+    m_avframe_device->width = args.width;
+    m_avframe_device->height = args.height;
     m_avframe_device->format = AV_PIX_FMT_YUV422P;
-    m_avframe_device->format = avDeviceFormat;
+    m_avframe_device->format = formats::get_av_pixel_format_from_string(args.av_device_format_str);
 
-    m_avframe_rgb->width = width;
-    m_avframe_rgb->height = height;
+    m_avframe_rgb->width = args.width;
+    m_avframe_rgb->height = args.height;
     m_avframe_rgb->format = AV_PIX_FMT_RGB24;
 
     m_sws_context = sws_getContext(
-      width, height, (AVPixelFormat)m_avframe_device->format,
-      width, height, (AVPixelFormat)m_avframe_rgb->format, SWS_FAST_BILINEAR,
+      args.width, args.height, (AVPixelFormat)m_avframe_device->format,
+      args.width, args.height, (AVPixelFormat)m_avframe_rgb->format, SWS_FAST_BILINEAR,
       NULL, NULL, NULL);
 
     // Suppress warnings from ffmpeg libraries to avoid spamming the console
@@ -122,8 +121,8 @@ public:
     av_log_set_flags(AV_LOG_SKIP_REPEATED);
     // av_log_set_flags(AV_LOG_PRINT_LEVEL);
 
-    m_avcodec_context->width = width;
-    m_avcodec_context->height = height;
+    m_avcodec_context->width = args.width;
+    m_avcodec_context->height = args.height;
     m_avcodec_context->pix_fmt = (AVPixelFormat)m_avframe_device->format;
     m_avcodec_context->codec_type = AVMEDIA_TYPE_VIDEO;
 

--- a/include/usb_cam/formats/mono.hpp
+++ b/include/usb_cam/formats/mono.hpp
@@ -44,7 +44,7 @@ namespace formats
 class MONO8 : public pixel_format_base
 {
 public:
-  MONO8()
+  explicit MONO8(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "mono8",
       V4L2_PIX_FMT_GREY,
@@ -52,14 +52,16 @@ public:
       1,
       8,
       false)
-  {}
+  {
+    (void)args;
+  }
 };
 
 
 class MONO16 : public pixel_format_base
 {
 public:
-  MONO16()
+  explicit MONO16(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "mono16",
       V4L2_PIX_FMT_Y16,
@@ -67,7 +69,9 @@ public:
       1,
       16,
       false)
-  {}
+  {
+    (void)args;
+  }
 };
 
 
@@ -75,7 +79,7 @@ public:
 class Y102MONO8 : public pixel_format_base
 {
 public:
-  explicit Y102MONO8(const int & number_of_pixels)
+  explicit Y102MONO8(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "y102mono8",
       V4L2_PIX_FMT_Y10,
@@ -83,7 +87,7 @@ public:
       1,
       8,
       true),
-    m_number_of_pixels(number_of_pixels)
+    m_number_of_pixels(args.pixels)
   {}
 
   /// @brief Convert a Y10 (MONO10) image to MONO8

--- a/include/usb_cam/formats/pixel_format_base.hpp
+++ b/include/usb_cam/formats/pixel_format_base.hpp
@@ -84,7 +84,7 @@ public:
 
   /// @brief String value of V4L2 capture pixel format
   /// @return std::string V4L2 capture pixel format
-  inline std::string v4l2_str() { return usb_cam::conversions::FCC2S(m_v4l2); }
+  inline std::string v4l2_str() {return usb_cam::conversions::FCC2S(m_v4l2);}
 
   /// @brief Name of output pixel (encoding) format to ROS
   /// @return

--- a/include/usb_cam/formats/pixel_format_base.hpp
+++ b/include/usb_cam/formats/pixel_format_base.hpp
@@ -43,6 +43,21 @@ namespace formats
 {
 
 
+/// @brief Helper structure to standardize all pixel_format_base
+/// constructors, so that they all have the same argument type.
+///
+/// This should also be easily extendable in the future if we need to add additional
+/// arguments for future pixel format(s) that are added.
+typedef struct
+{
+  std::string name = "";
+  int width = 640;
+  int height = 480;
+  size_t pixels = 640 * 480;
+  std::string av_device_format_str = "AV_PIX_FMT_YUV422P";
+} format_arguments_t;
+
+
 /// @brief Base pixel format class. Provide all necessary information for converting between
 /// V4L2 and ROS formats. Meant to be overridden if conversion function is required.
 class pixel_format_base

--- a/include/usb_cam/formats/pixel_format_base.hpp
+++ b/include/usb_cam/formats/pixel_format_base.hpp
@@ -35,7 +35,7 @@
 #include "linux/videodev2.h"
 
 #include "usb_cam/constants.hpp"
-
+#include "usb_cam/conversions.hpp"
 
 namespace usb_cam
 {
@@ -79,8 +79,12 @@ public:
   inline std::string name() {return m_name;}
 
   /// @brief Integer value of V4L2 capture pixel format
-  /// @return
+  /// @return uint32_t V4L2 capture pixel format
   inline uint32_t v4l2() {return m_v4l2;}
+
+  /// @brief String value of V4L2 capture pixel format
+  /// @return std::string V4L2 capture pixel format
+  inline std::string v4l2_str() { return usb_cam::conversions::FCC2S(m_v4l2); }
 
   /// @brief Name of output pixel (encoding) format to ROS
   /// @return

--- a/include/usb_cam/formats/rgb.hpp
+++ b/include/usb_cam/formats/rgb.hpp
@@ -44,7 +44,7 @@ namespace formats
 class RGB8 : public pixel_format_base
 {
 public:
-  RGB8()
+  explicit RGB8(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "rgb8",
       V4L2_PIX_FMT_RGB332,
@@ -52,7 +52,9 @@ public:
       3,
       8,
       false)
-  {}
+  {
+    (void)args;
+  }
 };
 
 }  // namespace formats

--- a/include/usb_cam/formats/uyvy.hpp
+++ b/include/usb_cam/formats/uyvy.hpp
@@ -44,7 +44,7 @@ namespace formats
 class UYVY : public pixel_format_base
 {
 public:
-  UYVY()
+  explicit UYVY(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "uyvy",
       V4L2_PIX_FMT_UYVY,
@@ -52,14 +52,16 @@ public:
       2,
       8,
       false)
-  {}
+  {
+    (void)args;
+  }
 };
 
 
 class UYVY2RGB : public pixel_format_base
 {
 public:
-  explicit UYVY2RGB(const int & number_of_pixels)
+  explicit UYVY2RGB(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "uyvy2rgb",
       V4L2_PIX_FMT_UYVY,
@@ -67,7 +69,7 @@ public:
       2,
       8,
       true),
-    m_number_of_pixels(number_of_pixels)
+    m_number_of_pixels(args.pixels)
   {}
 
   /// @brief In this format each four bytes is two pixels.

--- a/include/usb_cam/formats/yuyv.hpp
+++ b/include/usb_cam/formats/yuyv.hpp
@@ -44,7 +44,7 @@ namespace formats
 class YUYV : public pixel_format_base
 {
 public:
-  YUYV()
+  explicit YUYV(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "yuyv",
       V4L2_PIX_FMT_YUYV,
@@ -52,14 +52,16 @@ public:
       2,
       8,
       false)
-  {}
+  {
+    (void)args;
+  }
 };
 
 
 class YUYV2RGB : public pixel_format_base
 {
 public:
-  explicit YUYV2RGB(const int & number_of_pixels)
+  explicit YUYV2RGB(const format_arguments_t & args = format_arguments_t())
   : pixel_format_base(
       "yuyv2rgb",
       V4L2_PIX_FMT_YUYV,
@@ -67,7 +69,7 @@ public:
       3,
       8,
       true),
-    m_number_of_pixels(number_of_pixels)
+    m_number_of_pixels(args.pixels)
   {}
 
   /// @brief In this format each four bytes is two pixels.

--- a/include/usb_cam/usb_cam.hpp
+++ b/include/usb_cam/usb_cam.hpp
@@ -61,6 +61,40 @@ namespace usb_cam
 using usb_cam::utils::io_method_t;
 using usb_cam::formats::pixel_format_base;
 
+/// @brief Add more formats here and to driver_supported_formats below as
+/// they are added to this library
+using usb_cam::formats::RGB8;
+using usb_cam::formats::YUYV;
+using usb_cam::formats::YUYV2RGB;
+using usb_cam::formats::UYVY;
+using usb_cam::formats::UYVY2RGB;
+using usb_cam::formats::MONO8;
+using usb_cam::formats::MONO16;
+using usb_cam::formats::Y102MONO8;
+using usb_cam::formats::RAW_MJPEG;
+using usb_cam::formats::MJPEG2RGB;
+using usb_cam::formats::M4202RGB;
+
+
+/// @brief list all supported formats that this driver supports
+std::vector<std::shared_ptr<pixel_format_base>> driver_supported_formats(
+  const formats::format_arguments_t & args = formats::format_arguments_t())
+{
+  std::vector<std::shared_ptr<pixel_format_base>> fmts = {
+    std::make_shared<RGB8>(args),
+    std::make_shared<YUYV>(args),
+    std::make_shared<YUYV2RGB>(args),
+    std::make_shared<UYVY>(args),
+    std::make_shared<UYVY2RGB>(args),
+    std::make_shared<MONO8>(args),
+    std::make_shared<MONO16>(args),
+    std::make_shared<Y102MONO8>(args),
+    std::make_shared<RAW_MJPEG>(args),
+    std::make_shared<MJPEG2RGB>(args),
+    std::make_shared<M4202RGB>(args),
+  };
+  return fmts;
+}
 
 typedef struct
 {
@@ -273,6 +307,51 @@ public:
     return m_supported_formats;
   }
 
+  /// @brief Check if the given format is supported by this device
+  /// If it is supported, set the m_pixel_format variable to it.
+  /// @param format the format to check if it is supported
+  /// @return bool true if the given format is supported, false otherwise
+  inline bool set_pixel_format(const formats::format_arguments_t & args)
+  {
+    bool result = false;
+
+    std::shared_ptr<pixel_format_base> found_driver_format = nullptr;
+
+    // First check if given format is supported by this driver
+    for (auto driver_fmt : driver_supported_formats(args)) {
+      if (driver_fmt->name() == args.name) {
+        found_driver_format = driver_fmt;
+      }
+    }
+
+    if (found_driver_format == nullptr) {
+      // List the supported formats of this driver for the user before throwing
+      std::cerr << "This driver supports the following formats:" << std::endl;
+      for (auto driver_fmt : driver_supported_formats(args)) {
+        std::cerr << "\t" << driver_fmt->name() << std::endl;
+      }
+      throw std::invalid_argument(
+              "Specified format `" + args.name + "` is unsupported by this ROS driver"
+      );
+    }
+
+    std::cout << "This device supports the following formats:" << std::endl;
+    for (auto fmt : this->supported_formats()) {
+      // Always list the devices supported formats for the user
+      std::cout << "\t" << fmt.format.description << " ";
+      std::cout << fmt.v4l2_fmt.width << " x " << fmt.v4l2_fmt.height << " (";
+      std::cout << fmt.v4l2_fmt.discrete.denominator / fmt.v4l2_fmt.discrete.numerator << " Hz)";
+      std::cout << std::endl;
+
+      if (fmt.v4l2_fmt.pixel_format == found_driver_format->v4l2()) {
+        result = true;
+        m_image.pixel_format = found_driver_format;
+      }
+    }
+
+    return result;
+  }
+
   /// @brief Set pixel format from parameter list. Required to have logic within UsbCam object
   /// in case pixel format class requires additional information for conversion function
   /// (e.g. number of pixels, width, height, etc.)
@@ -280,50 +359,21 @@ public:
   /// @return pixel format structure corresponding to a given name
   inline std::shared_ptr<pixel_format_base> set_pixel_format(const parameters_t & parameters)
   {
-    using usb_cam::formats::RGB8;
-    using usb_cam::formats::YUYV;
-    using usb_cam::formats::YUYV2RGB;
-    using usb_cam::formats::UYVY;
-    using usb_cam::formats::UYVY2RGB;
-    using usb_cam::formats::MONO8;
-    using usb_cam::formats::MONO16;
-    using usb_cam::formats::Y102MONO8;
-    using usb_cam::formats::RAW_MJPEG;
-    using usb_cam::formats::MJPEG2RGB;
-    using usb_cam::formats::M4202RGB;
+    // create format arguments structure
+    formats::format_arguments_t format_args({
+        parameters.pixel_format_name,
+        parameters.image_width,
+        parameters.image_height,
+        m_image.number_of_pixels,
+        parameters.av_device_format,
+      });
 
-    if (parameters.pixel_format_name == "rgb8") {
-      m_image.pixel_format = std::make_shared<RGB8>();
-    } else if (parameters.pixel_format_name == "yuyv") {
-      m_image.pixel_format = std::make_shared<YUYV>();
-    } else if (parameters.pixel_format_name == "yuyv2rgb") {
-      // number of pixels required for conversion method
-      m_image.pixel_format = std::make_shared<YUYV2RGB>(m_image.number_of_pixels);
-    } else if (parameters.pixel_format_name == "uyvy") {
-      m_image.pixel_format = std::make_shared<UYVY>();
-    } else if (parameters.pixel_format_name == "uyvy2rgb") {
-      // number of pixels required for conversion method
-      m_image.pixel_format = std::make_shared<UYVY2RGB>(m_image.number_of_pixels);
-    } else if (parameters.pixel_format_name == "mjpeg") {
-      m_image.pixel_format = std::make_shared<RAW_MJPEG>(
-        formats::get_av_pixel_format_from_string(parameters.av_device_format));
-    } else if (parameters.pixel_format_name == "mjpeg2rgb") {
-      m_image.pixel_format = std::make_shared<MJPEG2RGB>(
-        m_image.width, m_image.height,
-        formats::get_av_pixel_format_from_string(parameters.av_device_format));
-    } else if (parameters.pixel_format_name == "m4202rgb") {
-      m_image.pixel_format = std::make_shared<M4202RGB>(
-        m_image.width, m_image.height);
-    } else if (parameters.pixel_format_name == "mono8") {
-      m_image.pixel_format = std::make_shared<MONO8>();
-    } else if (parameters.pixel_format_name == "mono16") {
-      m_image.pixel_format = std::make_shared<MONO16>();
-    } else if (parameters.pixel_format_name == "y102mono8") {
-      m_image.pixel_format = std::make_shared<Y102MONO8>(m_image.number_of_pixels);
-    } else {
+    // Look for specified pixel format
+    if (!this->set_pixel_format(format_args)) {
       throw std::invalid_argument(
-              "Unsupported pixel format specified: " +
-              parameters.pixel_format_name);
+              "Specified format `" + parameters.pixel_format_name + "` is unsupported by the " +
+              "selected device `" + parameters.device_name + "`"
+      );
     }
 
     return m_image.pixel_format;

--- a/include/usb_cam/utils.hpp
+++ b/include/usb_cam/utils.hpp
@@ -46,15 +46,12 @@ extern "C" {
 #include "linux/videodev2.h"
 
 #include "usb_cam/constants.hpp"
-#include "usb_cam/formats/pixel_format_base.hpp"
 
 
 namespace usb_cam
 {
 namespace utils
 {
-
-using usb_cam::formats::pixel_format_base;
 
 
 /// @brief Read more on IO methods here: https://lwn.net/Articles/240667/

--- a/src/ros2/usb_cam_node.cpp
+++ b/src/ros2/usb_cam_node.cpp
@@ -189,19 +189,9 @@ void UsbCamNode::init()
     rclcpp::shutdown();
     return;
   }
+
   // configure the camera
   m_camera->configure(m_parameters, io_method);
-
-  RCLCPP_INFO(this->get_logger(), "This devices supproted formats:");
-  for (auto fmt : m_camera->supported_formats()) {
-    RCLCPP_INFO(
-      this->get_logger(),
-      "\t%s: %d x %d (%d Hz)",
-      fmt.format.description,
-      fmt.v4l2_fmt.width,
-      fmt.v4l2_fmt.height,
-      fmt.v4l2_fmt.discrete.denominator / fmt.v4l2_fmt.discrete.numerator);
-  }
 
   set_v4l2_params();
 

